README.md
# IncreaseNumber

----

[![spm version](http://spmjs.io/badge/increasenumber)](http://spmjs.io/package/increasenumber)

数字自动增加的动画组件

- 支持参数修改动画时间
- 支持函数式输入控制变化进程
- 支持整数模式和小数模式
- 支持金额格式化


## Normal usage

```html
<div>
  <p class="data-increase" data-num="200000">0</p>
  <p class="data-increase" data-num="1000">0</p>
  <p><span class="data-increase data-percent" data-num="95.88">0</span> %</p>
</div>
```

```js
seajs.use('increasenumber', function(IncreaseNumber) {
  var increase = new IncreaseNumber('.data-increase');
  increase.start();
});
```

## API

### <span class="option-name">start()</span>

动画启动函数。

### <span class="option-name">stop()</span>

停止当前动画。

### <span class="option-name">setEvent(key, function)</span>

设置事件方法，目前支持<code>'beforeShow'</code>设置数字显示前的变化，例如：

```js
increase.setEvent('beforeShow', function(value, index, element){
	return value + '%';
});
```

### <span class="option-name">setAnimeFunction(function)</span>

设置数字变化的控制公式，与设置options中的animeFunc一致。

### <span class="option-name">IncreaseNumber.moneyFormat(num, n)</span>

金额格式化工具，输入num金额和保留小数点后n位，返回带逗号分割的标准金额格式。


## API (Options)

### <span class="option-name">decimalClass</span>

<span class="option-value">default: 'data-decimal'</span>

区分小数和整数的类名，存在该类名的数字默认显示两位小数，并且对应的param.isDecimal为true。

### <span class="option-name">dataAttr</span>

<span class="option-value">default: 'data-num'</span>

获取数据源的属性名。

### <span class="option-name">start</span>

<span class="option-value">default: 0</span>

动画起始时，数字起始值或以该百分比乘以数字得到起始值。

<code>start</code>值大于等1于时为起始值，<code>start</code>值小于1时为起始百分比。

### <span class="option-name">totalTime</span>

<span class="option-value">default: 3000</span>

总共动画耗费时间（毫秒）。

### <span class="option-name">intervalTime</span>

<span class="option-value">default: 40</span>

动画帧间隔时间（毫米）。

### <span class="option-name">animeFunc</span>

<span class="option-value">default: <code>function(x, a, index, element)</code></span>

数字变化的控制公式，x为单位坐标，a为变化速度，index为当前元素在组内下标，element为当前元素的DOM对象。

通过改变animeFunc可以控制动画变化的进程，例如使用平稳变化的线性函数：

```js
var increase = new IncreaseNumber('.data-increase', {
  animeFunc: function(x, a, index, element) {
    return a * x;
  }
});
increase.start();
```


HISTORY.md
# History

---

## 1.0.0

`new` It is the first version of increasenumber.


tests/increasenumber-spec.js
var expect = require('expect.js');
var IncreaseNumber = require('../IncreaseNumber');

describe('increasenumber', function() {

  it('increasenumber is existed', function() {
    expect(IncreaseNumber).to.be.ok();
    expect(IncreaseNumber.moneyFormat).to.be.ok();
    expect(IncreaseNumber.moneyFormat).to.be.a('function');
  });

  it('moneyFormat success', function() {
    expect(IncreaseNumber.moneyFormat(123456789, 0)).to.eql('123,456,789');
    expect(IncreaseNumber.moneyFormat(12345.6789, 2)).to.eql('12,345.68');
  });

});


_site/IncreaseNumber.js
var $ = require('jquery');

// 数据自动增加
function IncreaseNumber (selection, options) {
  this.group = '.data-increase';
  this.options = {
    decimalClass: 'data-decimal',
    dataAttr: 'data-num',
    start: 0, // 起始值或起始百分比
    totalTime: 3000,
    intervalTime: 40,
    animeFunc: function (x, a) {
      return a * x;
    }
  };
  this.event = {
    'beforeShow': function(value, i) {
      return this.param[i].isDecimal ? Number(value.toFixed(2)) : Math.round(value);
    }
  };

  this.group = selection || this.group;
  if ($.isPlainObject(options)) {
    $.extend(true, this.options, options);
  }
};

IncreaseNumber.prototype = {
  _step: function() {
    var that = this;
    this.progress += this.options.intervalTime / this.options.totalTime;

    if (this.progress >= 1) {
      this.stop();
      return false;
    }

    $(this.group).each(function (i, e) {
      var x = that.progress;
      var a = that.param[i].a;
      var y = Number(that.options.animeFunc(x, a, i, e));

      y += that.param[i].y1;

      y = that.event['beforeShow'].call(that, y, i, e);
      $(this).text(y);
    });
  },
  _dataInit: function() {
    var that = this;
    // 初始化
    this.progress = 0;
    this.param = [];

    // 参数初始化
    $(this.group).each(function () {
      // 整数或小数
      var isDecimal = $(this).hasClass(that.options.decimalClass);
      var max = Number($(this).attr(that.options.dataAttr));

      var y1 = 0, a = 1;
      // 起始值=max - 固定值
      if (that.options.start < 0) {
        y1 = max - that.options.start;
        a = Math.abs(that.options.start);
      }
      // 起始值=固定值
      else if (that.options.start >= 1) {
        if (max > that.options.start) {
          y1 = that.options.start;
          a = max - that.options.start;
        } else {
          // 转为起始值=max - 固定值
          y1 = max - that.options.start;
          a = Math.abs(that.options.start);
        }
      } 
      // 起始值=max * 比例
      else {
        y1 = that.options.start * max;
        if (isDecimal) {
          y1 = Number(y1.toFixed(2));
        } else {
          y1 = Math.round(y1);
        }
        a = max - y1;
      }

      that.param.push({
        max: max,
        a: a,
        y1: y1,
        isDecimal: isDecimal
      });

      $(this).text(y1);
    });
  },
  start: function() {
    var that = this;

    clearInterval(this.interval);
    this._dataInit();

    this.interval = setInterval(function() {
      that._step();
    }, this.options.intervalTime);
  },
  stop: function() {
    var that = this;

    clearInterval(this.interval);

    // 显示最终值
    $(this.group).each(function (i, e) {
      var max = that.param[i].max;

      max = that.event['beforeShow'].call(that, max, i, e);
      $(this).text(max);
    });
  },
  setEvent: function(key, func) {
    this.event[key] = func;
  },
  setAnimeFunction: function(func) {
    this.options.animeFunc = func;
  }
};

// 金额格式化
IncreaseNumber.moneyFormat = function (num, n) {
  num = String(num.toFixed(n));
  var re = /(-?\d+)(\d{3})/;
  while(re.test(num)) num = num.replace(re, "$1,$2");
  return num;
};

module.exports = IncreaseNumber;
